gsi/gridmap_eppn_callout/globus_gridmap_eppn.c

```diff
@@ -615,15 +616,36 @@ globus_gridmap_eppn_callout(
         context, subject, &found_identity, shared_user_cert, shared_user_chain, shared_exp);
     if(result == GLOBUS_SUCCESS)
     {
-        if(desired_identity && strcmp(found_identity, desired_identity) != 0)
+        if(desired_identity == NULL)
         {
-            GLOBUS_GRIDMAP_CALLOUT_ERROR(
-                result,
-                GLOBUS_GRIDMAP_CALLOUT_LOOKUP_FAILED,
-                ("Credentials specify id of %s, can not allow id of %s.\n",
-                 found_identity, desired_identity));
-            globus_free(found_identity);
-            goto error;
+            if(getenv("GLOBUS_EPPN_FORCE_GRIDMAP"))
+            {
+                tmp_identity = globus_libc_strdup(found_identity);
+                *found_identity = '\0';  /* reset string */
```

Comment:
Aren't you leaking `found_identity` here? I think you should instead call
`globus_free(found_identity); found_identity=NULL;`
Note that your next call `globus_gss_assist_gridmap(tmp_identity, &found_identity);` will actually make `found_identity==NULL`, so just free-ing would already be sufficient.

Reply:
@msalle my C skills are severly lacking, so i 'll change the code as you suggest (but i don't think it leaks this way)

Reply:
I think it did: the memory pointed to by `found_identity` would no longer have any pointers left after the call to `globus_gss_assist_gridmap()`.
Another improvement (would save you one memory allocation) would be to reorganise a bit and use `tmp_identity` for the result of `globus_gss_assist_gridmap()` and upon success put the result in the original `found_identity`. You can then also skip the free of the `tmp_identity` in the error block:

    if(getenv("GLOBUS_EPPN_FORCE_GRIDMAP"))
    {
        rc = globus_gss_assist_gridmap(found_identity, &tmp_identity);
        if(rc != 0)
        {
            GLOBUS_GRIDMAP_CALLOUT_ERROR(
                result,
                GLOBUS_GRIDMAP_CALLOUT_LOOKUP_FAILED,
                ("Could not map eppn %s (subject %s)\n", tmp_identity, subject));
            goto error;
        }
        globus_free(found_identity);
        found_identity=tmp_identity;
    }

Reply:
@msalle thanks. 
but the `globus_free(found_identity);` before the `goto error` should be a `globus_free(tmp_identity)`, no? (and the error message should use `found_identity` instead of `tmp_identity`)

Reply:
No, the `tmp_identity` hasn't been allocated inside the `globus_gss_assist_gridmap()` or otherwise is freed inside that function. I agree with the error message, sorry about that one.

Reply:
@msalle thanks for the explanation!

Reply:
@msalle out of curiosity, why do the `GLOBUS_GRIDMAP_CALLOUT_ERROR` use the `result` rather than the `rc` value after the `globus_gss_assist_gridmap` in the else block? i copied it here, but i realize only that is probably correct in the lower else block because there also result is false? and that i should use `rc` here?